sdk/storage/azure-storage-queue/src/main/java/com/azure/storage/queue/QueueAsyncClient.java

```
@@ -586,26 +638,81 @@ public HttpPipeline getHttpPipeline() {
+     * and other metadata about the enqueued message.
+     * @throws QueueStorageException If the queue doesn't exist or the {@code visibilityTimeout} or {@code timeToLive}
+     * are outside of the allowed limits.
+     */
+    @ServiceMethod(returns = ReturnType.SINGLE)
+    public Mono<Response<SendMessageResult>> sendMessageWithResponse(BinaryData message, Duration visibilityTimeout,
+                                                                     Duration timeToLive) {
+        try {
+            return withContext(context -> sendMessageWithResponse(message, visibilityTimeout, timeToLive, context));
+        } catch (RuntimeException ex) {
+            return monoError(logger, ex);
+        }
+    }
+
+    Mono<Response<SendMessageResult>> sendMessageWithResponse(BinaryData message, Duration visibilityTimeout,
                                                               Duration timeToLive, Context context) {
         Integer visibilityTimeoutInSeconds = (visibilityTimeout == null) ? null : (int) visibilityTimeout.getSeconds();
         Integer timeToLiveInSeconds = (timeToLive == null) ? null : (int) timeToLive.getSeconds();
-        QueueMessage message = new QueueMessage().setMessageText(messageText);
+        String messageText = encodeMessage(message);
+        QueueMessage queueMessage = new QueueMessage().setMessageText(messageText);
         context = context == null ? Context.NONE : context;
 
         return client.getMessages()
-            .enqueueWithResponseAsync(queueName, message, visibilityTimeoutInSeconds, timeToLiveInSeconds,
+            .enqueueWithResponseAsync(queueName, queueMessage, visibilityTimeoutInSeconds, timeToLiveInSeconds,
                 null, null,
                 context.addData(AZ_TRACING_NAMESPACE_KEY, STORAGE_TRACING_NAMESPACE_VALUE))
             .map(response -> new SimpleResponse<>(response, response.getValue().get(0)));
     }
 
+    private String encodeMessage(BinaryData message) {
```

Could we turn this into a `Mono<String>` and have it return `Mono.error` on an invalid input? This will prevent introducing another scenario where we throw instead of return an error in a reactive stream (reducing the likelihood of the try/catch blocks being used in caller APIs)